<?xml version="1.0"?>
<robot xmlns:xacro="http://www.ros.org/wiki/xacro" name="bookshelf_robot">
  <!-- Include macros, materials and gazebo propersies files -->
  <xacro:include filename="$(find knock_robot)/urdf/bookshelf_robot.gazebo" />
  <xacro:include filename="$(find knock_robot)/urdf/materials.xacro"/>
  <xacro:include filename="$(find knock_robot)/urdf/macros.xacro"/>
  <!-- Constants -->
  <xacro:property name="PI" value="3.1415926535897931"/>
  <!-- ########### -->
  <!-- Mobile base -->
  <!-- ########### -->
  <!-- Chassis link properties -->
  <xacro:property name="chassis_length" value="0.9" />
  <xacro:property name="chassis_width" value="0.5" />
  <xacro:property name="chassis_height" value="0.1" />
  <xacro:property name="chassis_mass" value="2.0"/>
  <!-- Wheel links properties -->
  <xacro:property name="wheel_length" value="0.08" />
  <xacro:property name="wheel_radius" value="0.2" />
  <xacro:property name="wheel_mass" value="0.2" />
  <!-- Caster leg link propersties -->
  <xacro:property name="caster_leg_length" value="0.05" />
  <xacro:property name="caster_leg_width" value="0.05" />
  <xacro:property name="caster_leg_height" value="0.1" />
  <xacro:property name="caster_leg_mass" value="0.1"/>
  <!-- Caster leg link properties -->
  <xacro:property name="caster_radius" value="0.04"/>
  <xacro:property name="caster_mass" value="0.1"/>
  <!-- ############ -->
  <!-- Robotic arm  -->
  <!-- ############ -->
  <!-- Arm base properties -->
  <xacro:property name="arm_base_radius" value="0.20"/>
  <xacro:property name="arm_base_length" value="0.05"/>
  <xacro:property name="arm_base_mass" value="1" />
  <!-- Torso properties -->
  <xacro:property name="torso_radius" value="0.05"/>
  <xacro:property name="torso_length" value="1.2"/>
  <xacro:property name="torso_mass" value="0.25" />
  <!-- Arm properties -->
  <xacro:property name="arm_radius" value="0.05"/>
  <xacro:property name="arm_length" value="0.5"/>
  <xacro:property name="arm_mass" value="0.05" />
  <!-- Elbow properties -->
  <xacro:property name="elbow_radius" value="0.05"/>
  <xacro:property name="elbow_length" value="0.7"/>
  <xacro:property name="elbow_mass" value="0.1" />
  <!-- Hand properties -->
  <!-- Hand base -->
  <xacro:property name="hand_base_length" value="0.08" />
  <xacro:property name="hand_base_radius" value="0.01" />
  <xacro:property name="hand_base_mass" value="0.05"/>
  <!-- Hand palm -->
  <xacro:property name="hand_palm_length" value="0.2" />
  <xacro:property name="hand_palm_width" value="0.01" />
  <xacro:property name="hand_palm_height" value="0.05" />
  <xacro:property name="hand_palm_mass" value="0.08"/>
  <!-- Hand finger -->
  <xacro:property name="hand_finger_length" value="0.2" />
  <xacro:property name="hand_finger_width" value="0.01" />
  <xacro:property name="hand_finger_height" value="0.06" />
  <xacro:property name="hand_finger_mass" value="0.05"/>
  <!-- Camera -->
  <xacro:property name="camera_dimensions" value="0.05" />
  <xacro:property name="camera_mass" value="0.01" />
  <!-- Laser Scanner -->

  <xacro:property name="lidar_base_length" value="0.04" />
  <xacro:property name="lidar_base_radius" value="0.01" />
  <xacro:property name="lidar_base_mass" value="0.05"/>

  <xacro:property name="lidar_length" value="0.05" />
  <xacro:property name="lidar_width" value="0.05" />
  <xacro:property name="lidar_height" value="0.05" />
  <xacro:property name="lidar_mass" value="0.1"/>


  
  <!-- ########################## -->
  <!-- Mobile base implementation -->
  <!-- ########################## -->
  <link name="footprint">
  </link>

  <joint name="base_joint" type="fixed">
    <parent link="footprint"/>
    <child link="chassis"/>
    <origin rpy="0 0 0" xyz="0 0 ${wheel_radius}"/>
  </joint>

  <link name="chassis">
    <visual>
      <geometry>
        <box size="${chassis_length} ${chassis_width} ${chassis_height}"/>
      </geometry>
      <material name="red"/>
    </visual>
    <collision>
      <geometry>
        <box size="${chassis_length} ${chassis_width} ${chassis_height}"/>
      </geometry>
    </collision>
    <inertial>
      <mass value="${chassis_mass}"/>
      <box_inertia m="${chassis_mass}" x="${chassis_length}" y="${chassis_width}" z="${chassis_height}"/>
    </inertial>
  </link>

  <link name="left_wheel">
    <visual>
      <geometry>
        <cylinder length="${wheel_length}" radius="${wheel_radius}"/>
      </geometry>
      <material name="blue"/>
    </visual>
    <collision>
      <geometry>
        <cylinder length="${wheel_length}" radius="${wheel_radius}"/>
      </geometry>
    </collision>
    <inertial>
      <mass value="${wheel_mass}"/>
      <cylinder_inertia m="${wheel_mass}" r="${wheel_radius}" h="${wheel_length}"/>
    </inertial>
  </link>

 <joint name="left_wheel_joint" type="continuous">
    <axis xyz="0 0 1"/>
    <parent link="chassis"/>
    <child link="left_wheel"/>
    <origin rpy="${-PI/2} 0 0" xyz="${-chassis_length/4} ${-chassis_width/2-wheel_length/2} 0}"/>
  </joint>

  <link name="right_wheel">
    <visual>
      <geometry>
        <cylinder length="${wheel_length}" radius="${wheel_radius}"/>
      </geometry>
      <material name="blue"/>
    </visual>
    <collision>
      <geometry>
        <cylinder length="${wheel_length}" radius="${wheel_radius}"/>
      </geometry>
    </collision>
    <inertial>
      <mass value="${wheel_mass}"/>
      <cylinder_inertia m="${wheel_mass}" r="${wheel_radius}" h="${wheel_length}"/>
    </inertial>
  </link>

  <joint name="right_wheel_joint" type="continuous">
    <axis xyz="0 0 1"/>
    <parent link="chassis"/>
    <child link="right_wheel"/>
    <origin rpy="${-PI/2} 0 0" xyz="${-chassis_length/4}  ${+chassis_width/2+wheel_length/2} 0"/>
  </joint>

  <link name="caster_leg">
    <visual>
      <geometry>
        <box size="${caster_leg_length} ${caster_leg_width} ${caster_leg_height}"/>
      </geometry>
      <material name="green"/>
    </visual>
    <collision>
      <geometry>
        <box size="${caster_leg_length} ${caster_leg_width} ${caster_leg_height}"/>
      </geometry>
    </collision>
    <inertial>
      <mass value="${caster_leg_mass}"/>
      <box_inertia m="${caster_leg_mass}" x="${caster_leg_length}" y="${caster_leg_width}" z="${caster_leg_height}"/>
    </inertial>
  </link>
  
  <joint name="chassis_caster_leg_joint" type="fixed">
    <parent link="chassis"/>
    <child link="caster_leg"/>
    <origin rpy="0 0 0" xyz="${3*chassis_length/8} 0 ${-chassis_height/2-caster_leg_height/2}"/>
  </joint>

  <link name="caster_wheel">
    <collision>
      <geometry>
        <sphere radius="${caster_radius}"/>
      </geometry>
    </collision>
    <visual>
      <geometry>
        <sphere radius="${caster_radius}"/>
      </geometry>
      <material name="black"/>
    </visual>
    <inertial>
      <mass value="${caster_mass}"/>
      <sphere_inertia m="${caster_mass}" r="${caster_radius}"/>
    </inertial>
  </link>

  <joint name="caster_leg_caster_wheel_joint" type="fixed">
    <parent link="caster_leg"/>
    <child link="caster_wheel"/>
    <origin rpy="0 0 0" xyz="0 0 ${-caster_radius/2-caster_leg_height/2}"/>
  </joint>
  <!-- Plugins -->
  <gazebo>
    <plugin name="differential_drive_controller" filename="libgazebo_ros_diff_drive.so">
      <leftJoint>left_wheel_joint</leftJoint>
      <rightJoint>right_wheel_joint</rightJoint>
      <robotBaseFrame>chassis</robotBaseFrame>
      <wheelSeparation>${chassis_width-wheel_length}</wheelSeparation>
      <wheelDiameter>${2*wheel_radius}</wheelDiameter>
      <odometryTopic>odom</odometryTopic>
      <odometryFrame>odom</odometryFrame>
      <publishWheelTF>false</publishWheelTF>
      <rosDebugLevel>na</rosDebugLevel>
      <publishWheelJointState>true</publishWheelJointState>
    </plugin>
  </gazebo>

  <!-- ########################## -->
  <!-- Robotic arm implementation -->
  <!-- ########################## -->

<link name="arm_base">
    <visual>
      <geometry>
        <cylinder length="${arm_base_length}" radius="${arm_base_radius}"/>
      </geometry>
      <material name="orange"/>
    </visual>
    <collision>
      <geometry>
        <cylinder length="${arm_base_length}" radius="${arm_base_radius}"/>
      </geometry>
    </collision>
    <inertial>
      <mass value="${arm_base_mass}"/>
      <cylinder_inertia m="${arm_base_mass}" r="${arm_base_radius}" h="${arm_base_length}"/>
    </inertial>
  </link>

  <joint name="chassis_arm_base" type="fixed">
    <parent link="chassis"/>
    <child link="arm_base"/>
    <origin rpy="${-PI} 0 0" xyz="0 0 ${chassis_height/2+arm_base_length/2}"/>
  </joint>

  <link name="torso">
    <visual>
      <geometry>
        <cylinder length="${torso_length}" radius="${torso_radius}"/>
      </geometry>
      <material name="orange"/>
    </visual>
    <collision>
      <geometry>
        <cylinder length="${torso_length}" radius="${torso_radius}"/>
      </geometry>
    </collision>
    <inertial>
      <mass value="${arm_base_mass}"/>
      <cylinder_inertia m="${arm_base_mass}" r="${torso_length}" h="${torso_radius}"/>
    </inertial>
  </link>

  <joint name="hip_joint" type="continuous">
    <axis xyz="0 0 1"/>
    <parent link="arm_base"/>
    <child link="torso"/>
    <origin rpy="0 0 0" xyz="0.0 0.0 ${-torso_length/2}"/>
    <!-- <limit effort="1" velocity="0.15" /> -->
  </joint>
  
  <link name="arm">
    <visual>
      <geometry>
        <cylinder length="${arm_length}" radius="${arm_radius}"/>
      </geometry>
      <origin rpy="0 0 0" xyz="0 0 ${-arm_length/2}"/>
      <material name="orange"/>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="0 0 ${-arm_length/2}"/>
      <geometry>
        <cylinder length="${arm_length}" radius="${arm_radius}"/>
      </geometry>
    </collision>
    <inertial>
      <origin rpy="0 0 0" xyz="0 0 ${-arm_length/2}"/>
      <mass value="${arm_mass}"/>
      <cylinder_inertia m="${arm_mass}" r="${arm_length}" h="${arm_radius}"/>
    </inertial>
  </link>

  <joint name="shoulder_joint" type="prismatic">
    <parent link="torso"/>
    <child link="arm"/>
    <origin rpy="0 ${-PI/2} 0" xyz="0 ${-2*arm_radius} ${-torso_length/2}"/>
    <axis xyz="1 0 0"/>
    <limit effort="100.0" lower="0.0" upper="1.1" velocity="100" />
  </joint>

  <link name="elbow">
    <visual>
      <geometry>
        <cylinder length="${elbow_length}" radius="${elbow_radius}"/>
      </geometry>
      <origin rpy="0 0 0" xyz="0 0 ${-elbow_length/2}"/>
      <material name="orange"/>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="0 0 ${-elbow_length/2}"/>
      <geometry>
        <cylinder length="${elbow_length}" radius="${elbow_radius}"/>
      </geometry>
    </collision>
    <inertial>
      <origin rpy="0 0 0" xyz="0 0 ${-arm_length/2}"/>
      <mass value="${elbow_mass}"/>
      <cylinder_inertia m="${elbow_mass}" r="${elbow_length}" h="${elbow_radius}"/>
    </inertial>
  </link>

  <joint name="elbow_joint" type="prismatic">
    <parent link="arm"/>
    <child link="elbow"/>
    <origin rpy="0 0 0" xyz="0 ${2*arm_radius} ${-arm_length/2}"/>
    <axis xyz="0 0 1"/>
    <limit effort="10.0" lower="0.0" upper="0.25" velocity="10" />
  </joint>

  <link name="hand_base">
    <visual>
      <geometry>
        <cylinder length="${hand_base_length}" radius="${hand_base_radius}"/>
      </geometry>
      <origin rpy="0 0 0" xyz="0 0 ${-hand_base_length/2}"/>
      <material name="red"/>
    </visual>
    <collision>
      <geometry>
        <cylinder length="${hand_base_length}" radius="${hand_base_radius}"/>
      </geometry>
      <origin rpy="0 0 0" xyz="0 0 0"/>
    </collision>
    <inertial>
      <mass value="${hand_base_mass}"/>
      <cylinder_inertia m="${hand_base_mass}" r="${hand_base_radius}" h="${hand_base_length}"/>
      <origin rpy="0 0 0" xyz="0 0 0"/>
    </inertial>
  </link>

  <joint name="wrist_joint" type="revolute">
    <parent link="elbow"/>
    <child link="hand_base"/>
    <axis xyz="0 1 0"/>
    <origin rpy="0 0 0" xyz="0 0 ${-elbow_length}-{hand_base_length/2}"/>
    <limit effort="1" lower="${-PI/2}" upper="${PI/2}" velocity="0.15" />
  </joint>

  <link name="hand_palm">
    <visual>
      <geometry>
        <box size="${hand_palm_length} ${hand_palm_width} ${hand_palm_height}"/>
      </geometry>
      <material name="blue"/>
    </visual>
    <collision>
      <geometry>
        <box size="${hand_palm_length} ${hand_palm_width} ${hand_palm_height}"/>
      </geometry>
    </collision>
    <inertial>
      <mass value="${hand_palm_mass}"/>
      <box_inertia m="${hand_palm_mass}" x="${hand_palm_length}" y="${hand_palm_width}" z="${hand_palm_height}"/>
    </inertial>
  </link>

  <joint name="base_palm_joint" type="fixed">
    <parent link="hand_base"/>
    <child link="hand_palm"/>
    <axis xyz="0 0 0"/>
    <origin rpy="${PI/2} 0 ${PI/2}" xyz="0 0 ${-hand_base_length}"/>
  </joint>

  <link name="hand_left_finger">
    <visual>
      <geometry>
        <box size="${hand_finger_length} ${hand_finger_width} ${hand_finger_height}"/>
      </geometry>
      <origin rpy="0 0 0" xyz="${hand_finger_length/2} 0 0"/>
      <material name="blue"/>
    </visual>
    <collision>
      <geometry>
        <box size="${hand_finger_length} ${hand_finger_width} ${hand_finger_height}"/>
      </geometry>
      <origin rpy="0 0 0" xyz="${hand_finger_length/2} 0 0"/>
    </collision>
    <inertial>
      <mass value="${hand_palm_mass}"/>
      <box_inertia m="${hand_palm_mass}" x="${hand_finger_length}" y="${hand_finger_width}" z="${hand_finger_height}"/>
      <origin rpy="0 0 0" xyz="${hand_finger_length/2} 0 0"/>
    </inertial>
  </link>

  <joint name="left_finger_joint" type="fixed">
    <parent link="hand_right_finger"/>
    <child link="hand_left_finger"/>
    <origin rpy="0 0 ${PI/2}" xyz="${-hand_palm_length} 0 0"/>
  </joint>

  <link name="hand_right_finger">
    <visual>
      <geometry>
        <box size="${hand_finger_length} ${hand_finger_width} ${hand_finger_height}"/>
      </geometry>
      <origin rpy="0 0 0" xyz="${-hand_finger_length/2} 0 0"/>
      <material name="blue"/>
    </visual>
    <collision>
      <geometry>
        <box size="${hand_finger_length} ${hand_finger_width} ${hand_finger_height}"/>
      </geometry>
      <origin rpy="0 0 0" xyz="${-hand_finger_length/2} 0 0"/>
    </collision>
    <inertial>
      <mass value="${hand_palm_mass}"/>
      <box_inertia m="${hand_palm_mass}" x="${hand_finger_length}" y="${hand_finger_width}" z="${hand_finger_height}"/>
      <origin rpy="0 0 0" xyz="${-hand_finger_length/2} 0 0"/>
    </inertial>
  </link>

  <joint name="right_finger_joint" type="fixed">
    <parent link="hand_palm"/>
    <child link="hand_right_finger"/>
    <origin rpy="0 0 ${PI/2}" xyz="${hand_palm_length/2} 0 0"/>
  </joint>

  <!-- Camera -->

  <link name="camera_link">
    <collision>
      <geometry>
      <box size="${camera_dimensions} ${camera_dimensions} ${camera_dimensions}"/>
      </geometry>
    </collision>
    <visual>
      <origin xyz="${-camera_dimensions/2} 0 0" rpy="0 0 0"/>
      <geometry>
      <box size="${camera_dimensions} ${camera_dimensions} ${camera_dimensions}"/>
      </geometry>
      <material name="white"/>
    </visual>
    <inertial>
      <mass value="${camera_mass}"/>
      <box_inertia m="${camera_mass}" x="${camera_dimensions}" y="${camera_dimensions}" z="${camera_dimensions}"/>
    </inertial>
  </link>

  <joint name="camera_joint" type="fixed">
    <origin xyz="${-elbow_radius} 0 ${-elbow_length}" rpy="${PI} ${PI/2} 0"/>
    <parent link="elbow"/>
    <child link="camera_link"/>
  </joint>

  <!-- Laser Scanner -->

<link name="lidar_base">
    <visual>
      <geometry>
        <cylinder length="${lidar_base_length}" radius="${lidar_base_radius}"/>
      </geometry>
      <origin rpy="0 0 0" xyz="0 0 ${lidar_base_length/2}"/>
      <material name="white"/>
    </visual>
    <collision>
      <geometry>
        <cylinder length="${lidar_base_length}" radius="${lidar_base_radius}"/>
      </geometry>
      <origin rpy="0 0 0" xyz="0 0 0"/>
    </collision>
    <inertial>
      <mass value="${lidar_base_mass}"/>
      <cylinder_inertia m="${lidar_base_mass}" r="${lidar_base_radius}" h="${lidar_base_length}"/>
      <origin rpy="0 0 0" xyz="0 0 0"/>
    </inertial>
  </link>

  <joint name="base_lidar_joint" type="fixed">
    <origin xyz="${chassis_length/3} 0 ${chassis_height/2}" rpy="0 0 0"/>
    <parent link="chassis"/>
    <child link="lidar_base"/>
  </joint>

  <link name="lidar_link">
    <visual>
      <geometry>
        <box size="${lidar_length} ${lidar_width} ${lidar_height}"/>
      </geometry>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <material name="white"/>
    </visual>
    <collision>
      <geometry>
        <box size="${lidar_length} ${lidar_width} ${lidar_height}"/>
      </geometry>
    </collision>
    <inertial>
      <mass value="${chassis_mass}"/>
      <box_inertia m="${lidar_mass}" x="${lidar_length}" y="${lidar_width}" z="${lidar_height}"/>
    </inertial>
  </link>

  <joint name="lidar_joint" type="fixed">
    <origin xyz="0 0 ${lidar_base_length}" rpy="0 0 0"/>
    <parent link="lidar_base"/>
    <child link="lidar_link"/>
  </joint>




  <transmission name="tran0">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="hip_joint">
      <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
    </joint>
    <actuator name="motor0">
      <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
  </transmission>
  <transmission name="tran1">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="shoulder_joint">
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
    </joint>
    <actuator name="motor1">
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
  </transmission>
  <transmission name="tran5">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="elbow_joint">
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
    </joint>
    <actuator name="motor5">
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
  </transmission>
  <transmission name="tran2">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="wrist_joint">
      <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
    </joint>
    <actuator name="motor2">
      <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
  </transmission>
  
  <gazebo>
    <plugin name="control" filename="libgazebo_ros_control.so"/>
  </gazebo>
  <gazebo>
    <plugin name="joint_state_publisher" filename="libgazebo_ros_joint_state_publisher.so">
      <jointName>hip_joint, shoulder_joint, elbow_joint, wrist_joint</jointName>
    </plugin>
  </gazebo>
</robot>